In [22]:
# # data ingestion
# import pandas as pd
# train_data = pd.read_csv('../data/avazu-ctr-prediction/train.gz',
#                             compression='gzip')
# train_data.head(3)

In [23]:
# train_data = train_data.sample(frac=1)
# train_data = train_data.head(8000)
# train_data.to_csv("../data/avazu-ctr-prediction/train.csv", index=False)

In [24]:
# test_data = pd.read_csv('../data/avazu-ctr-prediction/test.gz',
#                             compression='gzip')
# test_data = test_data.sample(frac=1)
# test_data = test_data.head(2000)
# test_data.to_csv("../data/avazu-ctr-prediction/test.csv", index=False)

# DATA INGESTION

In [25]:
import pandas as pd
train_df = pd.read_csv("../data/avazu-ctr-prediction/train.csv")
print('Data Loaded Successfully...')

Data Loaded Successfully...


# EXPLORATORY DATA ANALYSIS

In [26]:
train_df.info() # looking for the datatypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 24 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                8000 non-null   float64
 1   click             8000 non-null   int64  
 2   hour              8000 non-null   int64  
 3   C1                8000 non-null   int64  
 4   banner_pos        8000 non-null   int64  
 5   site_id           8000 non-null   object 
 6   site_domain       8000 non-null   object 
 7   site_category     8000 non-null   object 
 8   app_id            8000 non-null   object 
 9   app_domain        8000 non-null   object 
 10  app_category      8000 non-null   object 
 11  device_id         8000 non-null   object 
 12  device_ip         8000 non-null   object 
 13  device_model      8000 non-null   object 
 14  device_type       8000 non-null   int64  
 15  device_conn_type  8000 non-null   int64  
 16  C14               8000 non-null   int64  


In [27]:
train_df.describe() # statistical description

,id,click,hour,C1,banner_pos,device_type,device_conn_type,C14,C15,C16,C17,C18,C19,C20,C21
count,8.000000e+03,8000.000000,8.000000e+03,8000.00000,8000.000000,8000.000000,8000.000000,8000.000000,8000.000000,8000.000000,8000.000000,8000.000000,8000.000000,8000.000000,8000.000000
mean,9.262238e+18,0.164375,1.410256e+07,1004.98125,0.286625,1.023000,0.347500,18980.129375,319.209500,58.903000,2131.315625,1.432500,227.707000,52909.466125,84.497000
std,5.332532e+18,0.370639,2.979896e+02,1.10919,0.489133,0.546817,0.873263,4872.140729,22.112129,43.355156,596.530791,1.327371,349.719753,49978.958900,71.216397
min,1.084831e+16,0.000000,1.410210e+07,1001.00000,0.000000,0.000000,0.000000,375.000000,216.000000,36.000000,112.000000,0.000000,33.000000,-1.000000,13.000000
25%,4.660364e+18,0.000000,1.410230e+07,1005.00000,0.000000,1.000000,0.000000,17017.000000,320.000000,50.000000,1872.000000,0.000000,35.000000,-1.000000,23.000000
50%,9.218498e+18,0.000000,1.410260e+07,1005.00000,0.000000,1.000000,0.000000,20366.000000,320.000000,50.000000,2333.000000,2.000000,39.000000,100040.000000,61.000000
75%,1.384852e+19,0.000000,1.410281e+07,1005.00000,1.000000,1.000000,0.000000,22004.000000,320.000000,50.000000,2526.000000,3.000000,175.000000,100101.000000,111.000000
max,1.844488e+19,1.000000,1.410302e+07,1012.00000,7.000000,5.000000,5.000000,24043.000000,728.000000,480.000000,2757.000000,3.000000,1839.000000,100248.000000,253.000000


In [28]:
# checking for NaN values
train_df.isna().any()

id                  False
click               False
hour                False
C1                  False
banner_pos          False
site_id             False
site_domain         False
site_category       False
app_id              False
app_domain          False
app_category        False
device_id           False
device_ip           False
device_model        False
device_type         False
device_conn_type    False
C14                 False
C15                 False
C16                 False
C17                 False
C18                 False
C19                 False
C20                 False
C21                 False
dtype: bool

No NULL values in any of the columns, so we can proceed without fixing the NaN values

# FEATURE CLASSIFICATION



-  1. Time-based Features

- # `hour`

We can engineer:

- `hour_of_day`
- `day_of_week`
- `weekend_flag`

**Reason:**  
CTR changes heavily based on time and user activity patterns.



-  2. Ad Placement Features

- # `banner_pos`

Very important feature.

**Reason:**  
Ads shown higher on a page usually receive more clicks due to better visibility.



-  3. Site/App Context Features

- # Features:
- `site_category`
- `app_category`
- `site_domain`
- `app_domain`

**Reason:**  
User intent differs across websites and applications.

**Example:**  
Sports, news, and shopping platforms typically produce different CTR behaviors.



-  4. Device Features

- # Features:
- `device_type`
- `device_conn_type`
- `device_model`

**Reason:**  
CTR varies significantly across mobile, desktop, and tablet devices, as well as different network conditions.



-  5. High Cardinality Features

- # Features:
- `device_id`
- `device_ip`

**Reason:**  
These features may contain useful behavioral patterns but can also lead to overfitting because of extremely high cardinality.

**Approach:**  
- Frequency Encoding
- Hash Encoding
- Target Encoding
- Or removal after evaluation



-  6. Anonymous Categorical Features

- # Features:
- `C1`
- `C14` to `C21`

**Reason:**  
Although anonymized, these variables may encode important contextual or behavioral information useful for CTR prediction.

**Approach:**  
Retain initially and evaluate importance using feature importance or SHAP analysis.

In [29]:
# feature engineering on train-data
train_df["hour"] = train_df["hour"].astype(str) # Convert hour column to string
train_df["datetime"] = pd.to_datetime(train_df["hour"], format="%y%m%d%H") # Extract datetime from YYMMDDHH format

train_df["hour_of_day"] = train_df["datetime"].dt.hour # Hour of day
train_df["day_of_week"] = train_df["datetime"].dt.dayofweek # Day of week: Monday = 0, Sunday = 6
train_df["weekend_flag"] = train_df["day_of_week"].apply(lambda x: 1 if x >= 5 else 0) # Weekend flag

train_df[["hour", "datetime", "hour_of_day", "day_of_week", "weekend_flag"]].head() # Preview

,hour,datetime,hour_of_day,day_of_week,weekend_flag
0,14102912,2014-10-29 12:00:00,12,2,0
1,14102918,2014-10-29 18:00:00,18,2,0
2,14102205,2014-10-22 05:00:00,5,2,0
3,14102613,2014-10-26 13:00:00,13,6,1
4,14102604,2014-10-26 04:00:00,4,6,1


In [30]:
# feature engineering on test-data
test_df = pd.read_csv('../data/avazu-ctr-prediction/test.csv')
test_df["hour"] = test_df["hour"].astype(str) # Convert hour column to string
test_df["datetime"] = pd.to_datetime(test_df["hour"], format="%y%m%d%H") # Extract datetime from YYMMDDHH format

test_df["hour_of_day"] = test_df["datetime"].dt.hour # Hour of day
test_df["day_of_week"] = test_df["datetime"].dt.dayofweek # Day of week: Monday = 0, Sunday = 6
test_df["weekend_flag"] = test_df["day_of_week"].apply(lambda x: 1 if x >= 5 else 0) # Weekend flag

test_df[["hour", "datetime", "hour_of_day", "day_of_week", "weekend_flag"]].head() # Preview

,hour,datetime,hour_of_day,day_of_week,weekend_flag
0,14103107,2014-10-31 07:00:00,7,4,0
1,14103112,2014-10-31 12:00:00,12,4,0
2,14103116,2014-10-31 16:00:00,16,4,0
3,14103117,2014-10-31 17:00:00,17,4,0
4,14103106,2014-10-31 06:00:00,6,4,0


In [35]:
# feature selection
features = [
    'hour_of_day', 'day_of_week', 'weekend_flag',
    'banner_pos',
    'site_category', 'app_category', 'site_domain', 'app_domain',
    'device_type', 'device_conn_type', 'device_model',
    'C1', 'C14', 'C15', 'C16', 'C17', 'C18', 'C19', 'C20', 'C21']

X = train_df[features].copy()

# target column
y = train_df["click"]

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    test_size=0.2,
                                                    random_state=42)

In [36]:
from catboost import CatBoostClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score, log_loss, classification_report


cat_features = [
    'banner_pos',
    'site_category', 'app_category', 'site_domain', 'app_domain',
    'device_type', 'device_conn_type', 'device_model',
    'C1', 'C14', 'C15', 'C16', 'C17', 'C18', 'C19', 'C20', 'C21'
]

# CatBoost requires categorical columns as string
for col in cat_features:
    X_train[col] = X_train[col].astype(str)
    X_test[col] = X_test[col].astype(str)

In [37]:
model = CatBoostClassifier(
    iterations=500,
    learning_rate=0.05,
    depth=8,
    loss_function="Logloss",
    eval_metric="AUC",
    verbose=100,
    class_weights=[1, 5]
)
print('[INFO] Model Training Started...')

model.fit(
    X_train,
    y_train,
    cat_features=cat_features,
    eval_set=(X_test, y_test)
)
print('[INFO] Model Training Completed...')

y_pred_proba = model.predict_proba(X_test)[:, 1]
y_pred = model.predict(X_test)

print('\nPERFORMANCE METRICS')
print("ROC-AUC:", roc_auc_score(y_test, y_pred_proba))
print("Log Loss:", log_loss(y_test, y_pred_proba))
print(classification_report(y_test, y_pred))

[INFO] Model Training Started...
0:	test: 0.5962189	best: 0.5962189 (0)	total: 5.71ms	remaining: 2.85s
100:	test: 0.7043936	best: 0.7072602 (65)	total: 689ms	remaining: 2.72s
200:	test: 0.7062443	best: 0.7075669 (184)	total: 1.47s	remaining: 2.19s
300:	test: 0.7052793	best: 0.7075669 (184)	total: 2.32s	remaining: 1.53s
400:	test: 0.6930019	best: 0.7075669 (184)	total: 3.15s	remaining: 778ms
499:	test: 0.6942676	best: 0.7075669 (184)	total: 3.97s	remaining: 0us

bestTest = 0.7075669303
bestIteration = 184

Shrink model to first 185 iterations.
[INFO] Model Training Completed...

PERFORMANCE METRICS
ROC-AUC: 0.7076118148005183
Log Loss: 0.5820203157237076
              precision    recall  f1-score   support

           0       0.90      0.68      0.77      1353
           1       0.26      0.61      0.36       247

    accuracy                           0.67      1600
   macro avg       0.58      0.64      0.57      1600
weighted avg       0.80      0.67      0.71      1600



# Model Performance Evaluation

- ```ROC-AUC Score```

    ROC-AUC: 0.683

    The ROC-AUC score indicates that the model has moderate ability to distinguish between clicked and non-clicked ads.

    Since CTR prediction is a highly imbalanced classification problem, ROC-AUC is more informative than raw accuracy.

- ```Log Loss```

    Log Loss: 0.638

    Log Loss measures the quality of predicted probabilities.

    A lower Log Loss indicates better probability calibration and confidence estimation.

- ```Classification Report```

    - ```Class 0 (No Click)```

        - Precision: 0.91  
        - Recall: 0.56

        The model identifies non-click events reasonably well, although some non-click samples are misclassified as clicks.

    - ```Class 1 (Click)```

        - Precision: 0.24  
        - Recall: 0.71

        The model achieves high recall for clicked ads, meaning it successfully captures most click events.

        However, precision is relatively low, indicating false positives are present.

        This tradeoff is common in CTR prediction tasks where detecting potential clicks is usually prioritized over minimizing false positives.

- ```Accuracy```

    Accuracy: 58%

    Accuracy is not the primary metric for CTR prediction because the dataset is imbalanced.

    Metrics such as ROC-AUC, Recall, and Log Loss provide better insight into model quality.

- ```Overall Interpretation```

    The model demonstrates reasonable baseline CTR prediction capability with strong click-event recall and moderate ranking performance.

    Further improvements could include:

    - Hyperparameter tuning
    - Better categorical encoding
    - Feature interaction engineering
    - Using larger training data
    - Class imbalance optimization
    - SHAP-based feature analysis